# ğŸ˜€ ğŸŒ 2023.12.4

## ä¸€ã€principle of Analysis  

### Hi-C data processing

1. åŒç«¯æµ‹åºçš„å•ç»†èƒHi-Cå¤„ç†packageï¼š[NucProcess](https://github.com/TheLaueLab/nuc_processing)ã€‚
2. è¿™ä¸ªè½¯ä»¶ ([NucProcess](https://github.com/TheLaueLab/nuc_processing)) éœ€è¦åŸå§‹çš„ *FASTQ* æµ‹åºæ•°æ®, å‚è€ƒåŸºå› ç»„å’Œå®éªŒè¿‡ç¨‹ (é™åˆ¶æ€§å†…åˆ‡é…¶çš„ç§ç±»å’ŒDNAç‰‡æ®µå¤§å°çš„èŒƒå›´) . æ­¥éª¤å¦‚ä¸‹:
   - æŸ¥æ‰¾é™åˆ¶æ€§å†…åˆ‡é…¶ä½ç‚¹ï¼ˆrestriction enzyme cut sitesï¼‰
   - æ ¹æ®é…¶åˆ‡ä½ç‚¹åˆ†å‰²æ•°æ®
   - æ¯”å¯¹åˆ°å‚è€ƒåŸºå› ç»„ï¼ˆå»ºè®®ä½¿ç”¨Bowtie2ï¼‰ï¼ŒåŒæ—¶ä¸å…è®¸é”™é…ï¼ˆallowed for no base pair mismatchesï¼‰
   - æ¯”å¯¹ç»“æœåˆ†ä¸ºä¸¤éƒ¨åˆ†ï¼š1ï¼‰åŒç«¯æµ‹åºç»“æœéƒ½æ˜¯å”¯ä¸€åŒ¹é…ï¼›2ï¼‰å…¶ä¸­ä¸€æ¡æˆ–è€…ä¸¤æ¡æ¯”å¯¹ç»“æœå«ç³Šä¸æ¸…
   - The mapped read pairs were allocated to their originating chromosomal restriction enzyme fragment regions (i.e. one for each end) to identify the specific RE1-RE1 ligation junction involved. They were then filtered to remove pairs that represent obviously aberrant molecular events or contain no useful spatial information. The filtering stages are:
     - Removal of pairs that map to the same RE1 fragment: inward facing reads carry no ligation junction, useful or otherwise, and outward facing reads represent circularisation.
     - Removal of pairs that map to sequentially adjacent, inward facing RE1 fragments: these represent re-ligation and carry no useful structural information. 
     - Removal of pairs that map to RE1 fragments which are so close that they could be observed even if there is no ligation junction between them. 
     - Removal or pairs where the combined sequence separation from the read positions to the estimated ligation junction does not match the size of the fragments in the sequenced DNA library: i.e. the predicted length is too short and the sequenced molecule must contain unknown intermediate sequence, or the predicted length is too long and the sequenced molecule has somehow been truncated.
   - The redundancy in the restriction fragment pairs, where a particular contact is observed more than once due to sequence amplification, is removed. At the same time the data is filtered so that at least two separate, albeit often identical, molecules must be paired-end sequenced to confirm a ligation junction. Accordingly, the data is output in two separate groups: 1) redundant/supported pairs that are observed more than once, which are used in the structure calculations, and 2) unique/unsupported pairs that are only observed once, and which are not used in structure calculation. The latter have been shown to represent proportionately more noise. 
   - ç”±äºç ”ç©¶çš„ G1 å•å€ä½“ç»†èƒä¸­æ¯æ¡æŸ“è‰²ä½“åªæœ‰ä¸€ä¸ªæ‹·è´ï¼Œå› æ­¤ä»»ä½•æ‚ä¹±çš„é…å¯¹ï¼ˆå³é™åˆ¶æ€§ç‰‡æ®µæœ«ç«¯å‚ä¸äº†ä¸€æ¬¡ä»¥ä¸Šçš„æ¥è§¦ï¼‰éƒ½ä¼šè¢«åˆ é™¤ï¼ˆè¿™æ ·åˆ é™¤çš„contactséå¸¸å°‘ï¼‰ã€‚
   - è¾“å‡ºç»“æœä¸º`ncc`æ ¼å¼ï¼Œ`bin_size`é»˜è®¤å€¼ä¸º5 Mbã€‚



### 3D genome structure calculations and analysis 

3D genome structures were calculated using simulated annealing, of an initially random conformation of a particle-on-a-string representation of the chromosomes, to generate structures that were compatible with the experimental distance restraints derived from the Hi-C contacts. 

The software written to perform this task (which includes both a command line version and Jupyter notebook) is available at: https://github.com/TheLaueLab/nuc_dynamics. 



## äºŒã€ä»£ç éƒ¨åˆ†

~~~shell
# æ–‡ç« ä¸­çš„å•ç»†èƒHi-Cçš„nccæ ¼å¼æ–‡ä»¶åˆ©ç”¨nuc_toolså¯è§†åŒ–ä¸ºcontact map
(base) jialechen@Jiale-MacBook-Pro ~/Desktop/PhD/Pang_2022_GenomeBiol_3D/ncc $ nuc_tools contact_map  GSM6081054_Cell_ID_1.ncc -o Cell_ID_1.svg -sc # -scè¡¨ç¤ºä¸ºsingle cell

	INFO: Making PDF contact map for GSM6081054_Cell_ID_1.ncc
	INFO: Loading NCC format contact data
	INFO: Reading GSM6081054_Cell_ID_1.ncc
	INFO:  .. found 16,238 contacts
	
	INFO: Min. contig size not specified, using 10.0% of largest: 24,814,690 bp
	INFO: Considering 23 chromosomes/contigs
	INFO: Skipped 1 small chromosomes/contigs < 24,814,690 bp
	INFO: Binning contact lists
	INFO:  .. chrX - chrX : 20
	INFO: Full contact map size 1488 x 1488
	INFO:  .. making map Chromosome chrX (dpi=380)
	INFO: Written Cell_ID_1.pdf
~~~

<img src="/Users/jialechen/Desktop/PhD/Procedure_Circuit_Topology.assets/æˆªå±2023-12-04 16.56.39.png" alt="æˆªå±2023-12-04 16.56.39" style="zoom:33%;" />





# ğŸ˜‰ ğŸŒ§ï¸ 2023.12.5

## ä¸€ã€ä»£ç éƒ¨åˆ† ï¼ˆæ–‡ä»¶æ ¼å¼è½¬æ¢ï¼‰

~~~shell
# å°†å•ç»†èƒçš„nccæ ¼å¼æ–‡ä»¶å¯è§†åŒ–ä»£ç 
$ nuc_tools contact_map  GSM6081054_Cell_ID_1.ncc -o Cell_ID_1.svg -sc

# sample name 
Jiale-MacBook-Pro:ncc $ cat sample.txt
GSM6081054_Cell_ID_1
GSM6081055_Cell_ID_2
GSM6081056_Cell_ID_3
GSM6081057_Cell_ID_4
GSM6081058_Cell_ID_5
GSM6081059_Cell_ID_6
GSM6081060_Cell_ID_7
GSM6081061_Cell_ID_8
GSM6081062_Cell_ID_9
GSM6081063_Cell_ID_10
GSM6081064_Cell_ID_11
GSM6081065_Cell_ID_12
GSM6081066_Cell_ID_13
GSM6081067_Cell_ID_14
GSM6081068_Cell_ID_15
GSM6081069_Cell_ID_16
GSM6081070_Cell_ID_17
GSM6081071_Cell_ID_18
GSM6081072_Cell_ID_19
GSM6081073_Cell_ID_20
GSM6081074_Cell_ID_21
GSM6081075_Cell_ID_22
GSM6081076_Cell_ID_23
GSM6081077_Cell_ID_24
GSM6081078_Cell_ID_43
GSM6081079_Cell_ID_44
GSM6081080_Cell_ID_49
GSM6081081_Cell_ID_56
GSM6081082_Cell_ID_84
GSM6081083_Cell_ID_86
GSM6081084_Cell_ID_89
GSM6081085_Cell_ID_90
GSM6081086_Cell_ID_93

# æå–nccæ ¼å¼æ–‡ä»¶ä¸ºcontact txtæ ¼å¼
$ cat sample.txt|while read line ;do (nohup awk  '{print $1"\t"$2"\t"$7"\t"$8}' $line.ncc > $line.contact.txt &);done

-rw-r--r--@ 1 jialechen  staff   207K 12  5 11:13 GSM6081068_Cell_ID_15.contact.txt
-rw-r--r--@ 1 jialechen  staff   465K 12  5 11:13 GSM6081054_Cell_ID_1.contact.txt
-rw-r--r--@ 1 jialechen  staff   512K 12  5 11:13 GSM6081058_Cell_ID_5.contact.txt
-rw-r--r--@ 1 jialechen  staff   384K 12  5 11:13 GSM6081070_Cell_ID_17.contact.txt
-rw-r--r--@ 1 jialechen  staff   652K 12  5 11:13 GSM6081067_Cell_ID_14.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.3M 12  5 11:13 GSM6081061_Cell_ID_8.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.3M 12  5 11:13 GSM6081069_Cell_ID_16.contact.txt
-rw-r--r--@ 1 jialechen  staff   833K 12  5 11:13 GSM6081073_Cell_ID_20.contact.txt
-rw-r--r--@ 1 jialechen  staff   861K 12  5 11:13 GSM6081084_Cell_ID_89.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.7M 12  5 11:13 GSM6081056_Cell_ID_3.contact.txt
-rw-r--r--@ 1 jialechen  staff   872K 12  5 11:13 GSM6081081_Cell_ID_56.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.0M 12  5 11:13 GSM6081078_Cell_ID_43.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.0M 12  5 11:13 GSM6081083_Cell_ID_86.contact.txt
-rw-r--r--@ 1 jialechen  staff   803K 12  5 11:13 GSM6081086_Cell_ID_93.contact.txt
-rw-r--r--@ 1 jialechen  staff   783K 12  5 11:13 GSM6081085_Cell_ID_90.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.1M 12  5 11:13 GSM6081059_Cell_ID_6.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.4M 12  5 11:13 GSM6081079_Cell_ID_44.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.2M 12  5 11:13 GSM6081075_Cell_ID_22.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.4M 12  5 11:13 GSM6081080_Cell_ID_49.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.5M 12  5 11:13 GSM6081064_Cell_ID_11.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.3M 12  5 11:13 GSM6081060_Cell_ID_7.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.7M 12  5 11:13 GSM6081066_Cell_ID_13.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.7M 12  5 11:13 GSM6081082_Cell_ID_84.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.7M 12  5 11:13 GSM6081062_Cell_ID_9.contact.txt
-rw-r--r--@ 1 jialechen  staff   1.9M 12  5 11:13 GSM6081077_Cell_ID_24.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.5M 12  5 11:13 GSM6081063_Cell_ID_10.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.9M 12  5 11:13 GSM6081074_Cell_ID_21.contact.txt
-rw-r--r--@ 1 jialechen  staff   2.7M 12  5 11:13 GSM6081071_Cell_ID_18.contact.txt
-rw-r--r--@ 1 jialechen  staff   3.3M 12  5 11:13 GSM6081072_Cell_ID_19.contact.txt
-rw-r--r--@ 1 jialechen  staff   4.1M 12  5 11:13 GSM6081057_Cell_ID_4.contact.txt
-rw-r--r--@ 1 jialechen  staff   4.3M 12  5 11:13 GSM6081055_Cell_ID_2.contact.txt
-rw-r--r--@ 1 jialechen  staff   3.9M 12  5 11:13 GSM6081065_Cell_ID_12.contact.txt
-rw-r--r--@ 1 jialechen  staff   3.9M 12  5 11:13 GSM6081076_Cell_ID_23.contact.txt


# ç»™æ–‡ä»¶æ·»åŠ é¦–è¡Œ / chr_A	pos_A	chr_B	pos_B

$ cat sample.txt|while read line;do echo -e "chr_A\tpos_A\tchr_B\tpos_B"|cat - $line.contact.txt > $line.contact_pair.txt && mv $line.contact_pair.txt $line.contact.txt ;done

# æŒ‰ç…§å•ç»†èƒåœ¨/Users/jialechen/Desktop/PhD/CT/Pang_2022_GenomeBiol_3D/results/matricesè·¯å¾„ä¸‹æ–°å»ºæ–‡ä»¶å¤¹
$ cat sample.txt|sed 's/_/\t/' |cut -f 2 |while read line ;do mkdir -p ../results/matrices/$line ;done

~~~



# ğŸ¥± â˜ï¸ 2023.12.6

## ä¸€ã€ä»£ç åˆ†æ

- ä½¿ç”¨ExtractTopology_singlecellsHiC.ipynbä»£ç ä»å•ç»†èƒæ•°æ®ä¸­æå–CTç»“æœï¼Œåœ°å€ä¸ºï¼šhttp://localhost:8888/notebooks/Desktop/PhD/CT/Pang_2022_GenomeBiol_3D/Hi-C%20map%20analysis/ExtractTopology_singlecellsHiC.ipynbã€‚



# ğŸ˜ ğŸŒ 2022.12.7

## ä¸€ã€ä»£ç éƒ¨åˆ†

### 1. ä½¿ç”¨nuc_dynamicså°†nccæ–‡ä»¶ç”ŸæˆåŸºå› ç»„ç»“æ„pdbæ–‡ä»¶ï¼š

~~~shell
# Running NucDynamics
# Typical use, generating 10 conformational models in PDB format:
$ nuc_dynamics example_chromo_data/Cell_1_contacts.ncc -m 10 -f pdb

# Specifying the particle sizes (8 Mb, 2 Mb, 1 Mb, 500 kb) and an output file
name:
$ nuc_dynamics example_chromo_data/Cell_1_contacts.ncc -m 10 -f pdb -o Cell_1.pdb -s 8 2 1 0.5
~~~



ä½†æ˜¯ï¼ï¼æ–‡ç« ç»™çš„nccæ–‡ä»¶æ ¼å¼å¤ªè€ï¼Œç°æœ‰çš„å·¥å…·ä¸æ”¯æŒï¼Œæ— è¯­ã€‚è¿˜æ˜¯ç»§ç»­çœ‹Barbaraçš„ä»£ç ã€‚



# ğŸ˜„ â˜ï¸ 2023.12.11

## ä¸€ã€condensate project (preparation)

1. phase-separated biomolecular condesates
2. membraneless compartments (made of IDPs)
3. CellHesion device (atomic force microscope)
4. oppositely charged macromolecules: PLL:H and PLL:HS
5. laminin and Fus protein use various crowding agents and also various buffers



## äºŒã€analysis data

åˆ†æåµå·¢ç™Œï¼ˆovarian cancer cellï¼‰Hi-Cæ•°æ®ä¸­çš„ï¼š - P*ã€ŒEã€* / - S*ã€ŒUã€*/ - X*ã€ŒEã€*



# ğŸ¥± â˜ï¸ 2023.12.12



- æŠŠhic contact txtæå–Circuit topologyçš„ä»£ç ï¼ˆ/Users/jialechen/Desktop/PhD/CT/Pang_2022_GenomeBiol_3D/Hi-C map analysis/ExtractTopology_singlecellsHiC.ipynbï¼‰ä¸­åŠ å…¥æå–æŸ“è‰²ä½“ä¿¡æ¯çš„PSXã€‚



# ğŸ¥± â˜ï¸ 2023.12.13

- **laminin** structure and droplet formation 
- Laminins are large cell-adhesive glycoproteins
- Calculate the net charge





#  ğŸ˜‰ â˜ï¸ 2023.12.15

- Generate the hic.txt from ncc file 

~~~shell
$ cat sample.txt|while read line ;do (nohup awk  '{print $1"\t"$2"\t"$3"\t"$7"\t"$8"\t"$9"\t"$15}' $line.ncc > $line.hic.txt &);done

$ cat sample.txt|while read line;do echo -e "#chr_A\tstart_A\tend_A\tchr_B\tstart_B\tend_b\tnum_obs"|cat - $line.hic.txt > $line.hic_hder.txt  ;done

$ rm  *hic.txt

$ cat sample.txt|while read line;do mv $line.hic_hder.txt $line.hic.txt;done 
~~~

ç”Ÿæˆçš„ç»“æœä¸å°½äººæ„ï¼ŒRNAseqæ•°æ®æ˜¯tpmï¼Œæ˜¯æ ‡å‡†åŒ–çš„ï¼Œä½†æ˜¯æºä»£ç æ˜¯ç”¨çš„countså€¼ï¼Œæ‰€ä»¥å‡ºæ¥çš„ç»“æœå¤§éƒ¨åˆ†æ˜¯Nanã€‚



# ğŸ˜ƒ â˜€ï¸2024.02.12 å¤§å¹´åˆä¸‰

åŠå…¬å®¤é—¨é”åäº†ï¼Œåªèƒ½åœ¨æ—è¾¹çš„ä¼šè®®å®¤åŠå…¬ã€‚

ä¿®å¤äº†åŸæ¥coupleRNAseq_CT.ipynbçš„bugï¼Œåªæ˜¯åŠ äº†the codes belowï¼š

~~~shell
# Define the directory path
output_dir = 'results/correlation rna seq/{}/'.format(resolution_string)

# Create the directory if it doesn't exist
if not os.path.exists(output_dir):
    os.makedirs(output_dir)
~~~



# ğŸ˜„ â˜€ï¸ 2024.02.13

- Pang_2022_GenomeBiol_3D
  - Cell ID-1~24: single-cell Hi-C profiling of ovarian cancer cell line PEO1 and HEYA8
  - Cell ID-43/44/49/56/84/86/89/90/93: single-cell Hi-C profiling of OVCA429 shGRHL2 Tet-inducible, an ovarian cancer cell line 
- `contact map` transfer to `sparse matrix`

~~~python
## install dependent packages
!pip install numpy scipy scikit-image tqdm matplotlib

## ä½¿ç”¨skimageåº“ä¸­çš„ioæ¨¡å—æ‰“å¼€tiffæ–‡ä»¶ï¼Œå¹¶å°†å…¶è½¬æ¢ä¸ºç§‘å­¦è®¡ç®—åº“scipyä¸­çš„ç¨€ç–çŸ©é˜µæ ¼å¼csr_matrix	
from tqdm import tqdm
from skimage import io
from scipy.sparse import csr_matrix

# set path	
filename = "path/to/large_file.tiff"

# open and read tif 
image = io.imread(filename)

# å°†tifæ–‡ä»¶çŸ©é˜µè½¬æ¢ä¸ºç¨€ç–çŸ©é˜µ
sparse_matrix = csr_matrix(image)

## å¦‚æœ‰å¿…è¦ï¼Œå¯ä»¥é€‰æ‹©è¿›è¡Œé™é‡‡æ ·ä»¥å‡å°‘ç¨€ç–çŸ©é˜µçš„å¤§å°ã€‚å¯ä»¥ä½¿ç”¨skimageåº“ä¸­çš„transformæ¨¡å—è¿›è¡Œé™é‡‡æ ·ï¼š
from skimage import transform

# è®¾ç½®é™é‡‡æ ·å€æ•°
downsample = 5

# å¯¹ç¨€ç–çŸ©é˜µè¿›è¡Œé™é‡‡æ ·
downsampled_matrix = transform.downscale_local_mean(sparse_matrix.toarray(), (downsample, downsample))
~~~



# ğŸ˜… â˜”ï¸ 2024.02.14 

- Use `Local_matrix_analysis.ipynb` script to analyze the chromosome segments, and get the `.csv` file.
- This code can extract the `.tif` matrix and analyze the L-loops.



# 2024.02.15

å¼€å®Œstart up meetingï¼Œè€æ¿è®²äº†å¾ˆå¤šç‰©ç†çŸ¥è¯†ï¼ŒåŒ…æ‹¬å›ºä½“/æµä½“/ç‰›é¡¿æµä½“/viscoelasticç­‰ã€‚

å®‰æ’æˆ‘å°½å¿«æ‰¾åˆ°å¸¦æœ‰HBDçš„lamininçš„è›‹ç™½ï¼ˆå¸¦è§å…‰ï¼‰ç„¶åä¹°ï¼Œheparin binding domainä½äºlamininçš„alpha chain LG4 domainsã€‚



# 2024.02.16

- å‡†å¤‡åšå‘¨ä¸€çš„pptï¼Œæ±‡æŠ¥ä¸€ä¸‹ç»“æœï¼šé¦–å…ˆæ˜¯èƒŒæ™¯ï¼šä¸ºä»€ä¹ˆè¦åˆ†æå®«é¢ˆç™Œå•ç»†èƒHi-Cçš„æ•°æ®ï¼Œç»“æœï¼ˆhic contactå’Œct_mapï¼‰ã€‚



- å¾ªç¯è¯»å–hic.txtæ–‡ä»¶ä»£ç ï¼š

~~~python
import os

data_dir = '/Users/jialechen/Desktop/PhD/CT/Pang_2022_GenomeBiol_3D/data/'
file_names = [f for f in os.listdir(data_dir) if f.startswith('GSM') and f.endswith('.hic.txt')]


for file in file_names:
    contacts = pd.read_csv(os.path.join(data_dir, file), sep='\t')
~~~



# 2024.02.19

å•ç»†èƒæ•°æ®åˆ†æHGSOCï¼Œæ¨æ–­é€šè·¯ï¼Œæ‹Ÿæ—¶åºåˆ†æ

- ç‰›é¡¿æµä½“
- Hook solid



# 2024.02.20

- create the STAR genome index of repeats

  ~~~shell
  $ STAR --runThreadN 10 --runMode genomeGenerate \
   			 --genomeDir  ./ #folder of index results \
         --genomeFastaFiles ~/Desktop/Analysis_work/reference/Hsapiens/hg38.fa  \
         --sjdbGTFfile ~/Desktop/Analysis_work/reference/Hsapiens/repeat.hg38.gtf \
         --limitSjdbInsertNsj 5005834
  ~~~
  



# 2024.02.21 learn the concentration

learn the concentration with Dirk-jan and calcalate the different units about gram, Liter, etc.

| è‹±æ–‡åç§°     | ç¼©å†™  | ä¸­æ–‡åç§°     | å¤‡æ³¨                                                         |
| ------------ | ----- | ------------ | ------------------------------------------------------------ |
| gram         | g.    | å…‹           | é€‚åº”äºè¡¨ç¤ºé‡é‡çš„æƒ…å†µï¼Œå¦‚ï¼šthis chip is 10g by weight.        |
| centigram    | cg    | å˜å…‹         | é‡é‡å•ä½                                                     |
| milligram    | mmg   | æ¯«å…‹         | é‡é‡å•ä½                                                     |
| kilogram     | kg.   | åƒå…‹ï¼ˆå…¬æ–¤ï¼‰ | é€šå¸¸ï¼Œkiloè¿™ä¸ªè¯æ ¹è¡¨ç¤ºâ€œåƒâ€ï¼Œå¦‚kilogramï¼ˆåƒå…‹ï¼‰ï¼Œkilometer(åƒç±³)ã€‚æ­¤å¤„ï¼Œä¹Ÿæ˜¯ä¸€ä¸ªé‡é‡å•ä½ã€‚ |
| metric ton   | m.t.  | å…¬å¨         | é‡é‡å•ä½                                                     |
| long ton     | l.t.  | é•¿å¨         | é‡é‡å•ä½                                                     |
| short ton    | sh.t. | çŸ­å¨         | é‡é‡å•ä½                                                     |
| pound        | lb.   | ç£…           | é‡é‡å•ä½1lb.=0.454kg                                         |
| ounce        | oz    | ç›å¸         | é‡é‡å•ä½1 oz=31.1035g                                        |
| meter        | m     | ç±³           | é•¿åº¦å•ä½                                                     |
| kilometer    | km    | åƒç±³ï¼ˆå…¬é‡Œï¼‰ | é•¿åº¦å•ä½                                                     |
| decimeter    |       | åˆ†ç±³ï¼ˆå…¬å¯¸ï¼‰ | é•¿åº¦å•ä½                                                     |
| centimeter   | cm    | å˜ç±³ï¼ˆå…¬åˆ†ï¼‰ | é•¿åº¦å•ä½                                                     |
| millimeter   | mm    | æ¯«ç±³         | é•¿åº¦å•ä½                                                     |
| micrometer   | Î¼m    | å¾®ç±³         | é•¿åº¦å•ä½                                                     |
| nanometer    | nm    | çº³ç±³         | é•¿åº¦å•ä½                                                     |
| yard         | yd    | ç            | é•¿åº¦å•ä½                                                     |
| foot         | ft    | è‹±å°º         | é•¿åº¦å•ä½1ft=12in=30.48cm                                     |
| inch         | in.   | è‹±å¯¸         | é•¿åº¦å•ä½                                                     |
| square meter | sq.m  | å¹³æ–¹ç±³       | é¢ç§¯å•ä½                                                     |
| square foot  | sq.ft | å¹³æ–¹è‹±å°º     | é¢ç§¯å•ä½                                                     |
| square inch  | sq.in | å¹³æ–¹è‹±å¯¸     | é¢ç§¯å•ä½                                                     |
| square yard  | sq.yd | å¹³æ–¹ç        | é¢ç§¯å•ä½                                                     |
| liter        | l.    | å‡           | å®¹ç§¯å•ä½                                                     |
| milliliter   | ml    | æ¯«å‡         | å®¹ç§¯å•ä½                                                     |
| gallon       | gal   | åŠ ä»‘         | å®¹ç§¯å•ä½                                                     |
| pint         |       | å“è„±         | å®¹ç§¯å•ä½                                                     |
| bushel       | bu    | è’²å¼è€³       | å®¹ç§¯å•ä½                                                     |
| cubic meter  | cu.m  | ç«‹æ–¹ç±³       | ä½“ç§¯å•ä½                                                     |
| cubic foot   | cu.ft | ç«‹æ–¹è‹±å°º     | ä½“ç§¯å•ä½                                                     |
| piece        | pc    | åªï¼Œä»¶ï¼Œå—   | æ•°é‡å•ä½                                                     |
| package      | pkg   | ä»¶ï¼ŒåŒ…       | æ•°é‡å•ä½                                                     |
| pair         |       | åŒï¼Œå¯¹       | æ•°é‡å•ä½                                                     |
| set          |       | å°ï¼Œå¥—ï¼Œæ¶   | æ•°é‡å•ä½                                                     |
| dozen        | doz.  | æ‰“           | æ•°é‡å•ä½                                                     |
| gross        | gr.   | ç½—           | æ•°é‡å•ä½                                                     |
| great gross  | g.gr. | å¤§ç½—         | æ•°é‡å•ä½                                                     |
| ream         | rm    | ä»¤           | æ•°é‡å•ä½                                                     |
| roll         |       | å·           | æ•°é‡å•ä½                                                     |
| unit         |       | ä»¶ï¼Œè¾†       | æ•°é‡å•ä½                                                     |
| head         |       | å¤´           | æ•°é‡å•ä½                                                     |
| barrel       |       | æ¡¶           | æ•°é‡å•ä½                                                     |
| bag          |       | è¢‹           | æ•°é‡å•ä½                                                     |
| lot          |       | æ‰¹           | æ•°é‡å•ä½                                                     |
| bar          |       | æ¡           | æ•°é‡å•ä½                                                     |
| batch        |       | æ‰¹           | æ•°é‡å•ä½                                                     |
| bolt         |       | åŒ¹           | æ•°é‡å•ä½                                                     |
| sheet        |       | å¼            | æ•°é‡å•ä½                                                     |
| bunch        |       | ä¸²ï¼ŒæŸ       | æ•°é‡å•ä½                                                     |
| pile         |       | å †           | æ•°é‡å•ä½                                                     |
| group        |       | ç»„ï¼Œå¥—       | æ•°é‡å•ä½                                                     |
| portion      |       | ä»½           | æ•°é‡å•ä½                                                     |
| set          |       | å¥—           | æ•°é‡å•ä½                                                     |
| hour         | h     | å°æ—¶         | æ—¶é—´å•ä½                                                     |
| minute       | min   | åˆ†é’Ÿ         | æ—¶é—´å•ä½                                                     |
| second       | sec   | ç§’           | æ—¶é—´å•ä½                                                     |
| ampere       | A     | å®‰åŸ¹         | ç”µæµå•ä½                                                     |
| milliampere  | mA    | æ¯«å®‰         | ç”µæµå•ä½                                                     |
| microampere  | Î¼A    | å¾®å®‰         | ç”µæµå•ä½                                                     |
| volt         | V     | ä¼ç‰¹         | ç”µå‹å•ä½                                                     |
| millivolt    | mV    | æ¯«å¾®         | ç”µå‹å•ä½                                                     |
| microvolt    | Î¼V    | å¾®ä¼         | ç”µå‹å•ä½                                                     |
| Ohm          | Î©     | æ¬§å§†         | ç”µé˜»å•ä½                                                     |
| milliohm     | mÎ©    | æ¯«æ¬§         | ç”µé˜»å•ä½                                                     |
| microohm     | Î¼Î©    | å¾®æ¬§         | ç”µé˜»å•ä½                                                     |
| watt         | W     | ç“¦ç‰¹         | åŠŸç‡å•ä½                                                     |
| kilowatt     | KW    | åƒç“¦         | åŠŸç‡å•ä½                                                     |
| Kilocalorie  | kcal  | åƒå¡ï¼Œå¤§å¡   | çƒ­é‡å•ä½                                                     |
| calorie      | cal   | å¡è·¯é‡Œ       | çƒ­é‡å•ä½                                                     |



# 2024.02.22

- download the reference genome:https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/ with the gene name

- create STAR genome index 

  ~~~
  # coding_genes
  STAR --runThreadN 6 --runMode genomeGenerate --genomeDir coding_genes/ --genomeFastaFiles ~/Desktop/Analysis_work/reference/Hsapiens/hg38.fa --sjdbGTFfile ~/Desktop/Analysis_work/reference/Hsapiens/hg38.refGene.gtf --sjdbOverhang 149
  
  # repeats
  STAR --runThreadN 6 --runMode genomeGenerate --genomeDir repeats/ --genomeFastaFiles ~/Desktop/Analysis_work/reference/Hsapiens/hg38.fa --sjdbGTFfile ~/Desktop/Analysis_work/reference/Hsapiens/repeat.hg38.gtf --sjdbOverhang 149
  ~~~

  



# 2024.03.05

- remote connect lab computer and I can use it now : )
- I have downloaded the raw fastq data of the single cell data from Pang 2022 Genome Biology.
- try to use the `nuc_process` to process the raw data

~~~shell
# nuc_process

# step 0: change the fastq name from _1.fastq to _r_1.fastq
# The splitFastqBarcodes.py script is provided to split FASTQ files that represent many cells, each with a different barcode sequence, into separate paired read files. The script can be run as follows, specifying the names of the two paired, multiplex FASTQ read files after the script:
 
(base) msbb@msbb-System-Product-Name:~/jiale/projects/Pang_2022_GenomeBiol_3D/data/fastq$ cat /home/msbb/jiale/projects/Pang_2022_GenomeBiol_3D/data/fastq/sample.txt  |while read line ;do split_fastq_barcodes ${line}_r_1.fastq.gz ${line}_r_2.fastq.gz -s 3 ;done

   # INFO: No -b barcode file specified.
   # INFO: No FASTQ files will be written.
   # INFO: Only analysis will be performed.


# step 1: creates a genome index and RE-digest files:
(jiale) msbb@msbb-System-Product-Name:~/jiale/projects/Pang_2022_GenomeBiol_3D/data/fastq$ nuc_process -f ~/jiale/library/chromosome/*.fa -o cell_id_1 -v -a -k -re1 MboI -re2 AluI -s 150-2000 -n 12 -g ~/jiale/tools/nuc_processing/hg38/  FASTQ_FILE ./Cell_ID-1_r_1.fastq.gz ./Cell_ID-1_r_2.fastq.gz

~~~



# 2024.03.06 ğŸŒ & â˜€ï¸

â¡ï¸ make condensate with certain concentration: 

- HNa2O4P  100mM
- KCl              200mM
- PEG             230mg/ml
- BSA             30mg/ml

â¡ï¸ the density of the condensate is high.









# 2024.03.11

## ç»˜åˆ¶condensateä¸åŒæµ“åº¦ç»“æœå›¾

~~~python
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from scipy.interpolate import interp1d

# ä½ çš„æ•°æ®
data = {
    'Type': ['one', 'one', 'two', 'two', 'two', 'one', 'one', 'two', 'two', 'two', 'one', 'one', 'two', 'two'],
    'BSA_Concentration': [15, 15, 15, 15, 15, 30, 30, 30, 30, 30, 60, 60, 60, 60],
    'PEG_Concentration': [50, 100, 150, 200, 230, 50, 100, 150, 200, 230, 25, 50, 100, 150]
}

df = pd.DataFrame(data)

# åˆ†ç¦»ä¸¤ä¸ªç›¸åŒºçš„æ•°æ®
one_phase = df[df['Type'] == 'one']
two_phase = df[df['Type'] == 'two']

# è®¡ç®—æ¯ä¸ªBSAæµ“åº¦çº§åˆ«çš„"one"å’Œ"two"ç›¸åŒºPEGæµ“åº¦çš„å¹³å‡å€¼çš„ä¸­ç‚¹
bsa_levels = sorted(df['BSA_Concentration'].unique())
middle_points = []

for level in bsa_levels:
    one_avg = one_phase[one_phase['BSA_Concentration'] == level]['PEG_Concentration'].mean()
    two_avg = two_phase[two_phase['BSA_Concentration'] == level]['PEG_Concentration'].mean()
    middle_points.append((level, (one_avg + two_avg) / 2))

middle_points = np.array(middle_points)

# çº¿æ€§æ’å€¼æ¥ç”Ÿæˆå¹³æ»‘çš„åˆ†ç•Œçº¿
f_linear = interp1d(middle_points[:,0], middle_points[:,1], kind='linear', fill_value='extrapolate')
xnew = np.linspace(np.min(bsa_levels), np.max(bsa_levels), num=100, endpoint=True)

# ç»˜åˆ¶åŸå§‹æ•°æ®ç‚¹
plt.scatter(one_phase['BSA_Concentration'], one_phase['PEG_Concentration'], color='blue', label='One Phase')
plt.scatter(two_phase['BSA_Concentration'], two_phase['PEG_Concentration'], color='red', label='Two Phases')

# ç»˜åˆ¶å¹³æ»‘çš„åˆ†ç•Œçº¿
plt.plot(xnew, f_linear(xnew), 'k--', label='Phase Boundary')

# å¡«å……ä¸¤ä¸ªç›¸åŒºåŸŸ
plt.fill_between(xnew, f_linear(xnew), df['PEG_Concentration'].min() - 10, color='lightblue', alpha=0.2,label='One Phase')
plt.fill_between(xnew, f_linear(xnew), df['PEG_Concentration'].max() + 10, color='salmon', alpha=0.2,label='Two Phases')

plt.text(one_phase['BSA_Concentration'].mean(), one_phase['PEG_Concentration'].mean(), 'One Phase', color='blue', horizontalalignment='center', verticalalignment='bottom')
plt.text(two_phase['BSA_Concentration'].mean(), two_phase['PEG_Concentration'].mean(), 'Two Phases', color='red', horizontalalignment='center', verticalalignment='top')


plt.title('PEG-BSA Phase Diagram')
plt.xlabel('BSA  (mg/mL)')
plt.ylabel('PEG  (mg/mL)')
#plt.legend()
plt.show()

~~~



## è®¢è´­peptides and DNA ï¼ˆ2024-Science Advï¼‰

- æ–‡ç« æ¨èï¼šWe used ***poly(dT) sequences*** featuring different numbers of nucleotides ranging from 20 to 200 nt (dT20, dT40, dT90, and dT200).

| System        | ğ¸ à®º(RT) | ğ¸ à®º(RT) | ğ¸ à®º(RT) | ğ¸ à®ºàµ‡ âˆ†ğ¸ à®º(RT) |
| ------------- | ------- | ------- | ------- | ------------- |
| [RPRPP]5-dT40 | 10.35   | 8.44    | 8.72    | 9Â±1           |
| [RGRGG]5-dT40 | 15.97   | 18.87   | 17.23   | 17 Â± 1        |
| [RGYGG]3-dT40 | 27.01   | 22.07   | 21.58   | 24 Â± 3        |



# 2024.3.12 â˜ï¸

- peptides and DNAï¼šæ–‡ç« ä¸­ä½¿ç”¨æœ€å¤šçš„å°±æ˜¯[RGRGG]5-dT40ï¼Œæ‰€ä»¥æœ¬å®éªŒä¹Ÿå¯ä»¥ä½¿ç”¨åŒä¸Šåºåˆ—ã€‚
- ç”¨nuc_processå°†fastqè½¬ä¸ºnccï¼ˆnewï¼‰ï¼š

~~~shell
# github
$ nuc_process  /data/hi-c/Cell1.r[12].fq.gz -o Cell1 -v -a -k -re1 MboI -re2 AluI -s 50-5000 -n 8 -f /data/genome/mm10_chromosomes/*.fa -g /data/genome/mm10 -cn /data/genome/mm10_chromo_names.tsv

#my server
### Typical first-time use, which creates a genome index and RE-digest files:
$ nuc_process  -f /home/msbb/jiale/library/hg38.fa -o cell -v -a -k -re1 MboI -re2 AluI -s 150-2000 -n 12 -g /home/msbb/jiale/library/Bowtie2Index/hg38  ~/jiale/projects/Pang_2022_GenomeBiol_3D/data/fastq/Cell_ID-10_r[12].fastq

$ nuc_process Cell_ID-11_r_[12].fastq.gz  -o Cell1 -v -a -k -re1 MboI -re2 AluI -s 50-5000 -n 8  -f /home/msbb/jiale/library/hg38.fa  -g /home/msbb/jiale/library/Bowtie2Index/hg38  -cn ~/jiale/tools/nuc_processing/hg38_chromo_names.tsv
~~~



- Activate lion env: 

~~~
source venv/bin/activate
conda activate jiale 
~~~



# 2024.3.13

~~~bash
# use `nuc_sequence_names` to generate the tsv file

$ nuc_sequence_names  hg38_chromo_names.tsv j.chen@lacdr.leidenuniv.nl ~/jiale/library/hg38.fa
~~~



# 2024.3.14 â˜€ï¸

- Nuc_processè¿è¡ŒæˆåŠŸï¼ï¼ï¼

```bash
(jiale) msbb@msbb-System-Product-Name:/media/msbb/J.Chen_MSBB/PhD/CT/Pang_2022_GenomeBiol_3D/data/fastq$ nuc_process Cell_ID-10_r_[12].fastq -o Cell1 -v -a -k -re1 MboI -re2 AluI -s 50-5000 -n 8 -f /home/msbb/jiale/library/hg38.fa -g /home/msbb/jiale/library/Bowtie2Index/hg38 -cn ~/jiale/tools/nuc_processing/hg38_chromo_names.tsv


```

- change a loop command now

```bash
cat sample.txt|while read line ;do (nohup nuc_process ${line}_r_[12].fastq.gz -o $line -v -a -k -re1 MboI -re2 AluI -s 50-5000 -n 8 -f /home/msbb/jiale/library/hg38.fa -g /home/msbb/jiale/library/Bowtie2Index/hg38 -cn ~/jiale/tools/nuc_processing/hg38_chromo_names.tsv &);done 

```



- Use `nuc_dynmics` to transfer data from `ncc` to `pdb`

```bash
## Typical use, generating 10 conformational models in PDB format:

$ nuc_dynamics Cell_ID-10_nuc_pair.ncc -m 10 -f pdb

## Specifying the particle sizes (8 Mb, 2 Mb, 1 Mb, 500 kb) and an output file
name:

$ nuc_dynamics example_chromo_data/Cell_1_contacts.ncc -m 10 -f pdb -o Cell_1.pdb -s 8 2 1 0.5
```





# 2024.3.20

## transfer the hic contact files to bed files (same chromosome).



- ~~~bash
  $ head GSM6081054_Cell_ID_01.contact.txt                                               
  chr_A	pos_A	chr_B	pos_B
  chr10	100160223	chr2	67288359
  chr10	100201623	chr3	127643636
  chr10	100288384	chr2	225501213
  chr10	100288397	chr2	225501213
  chr10	10031568	chr2	24049144
  chr10	100322067	chr6	90732008
  chr10	100322078	chr6	90732008
  chr10	100331209	chr12	31235800
  chr10	100427285	chr5	63196830
  
  $ head Cell_ID_01.bed                                                                
  chr1	629508	629357	chr1	634058	634209
  chr1	2045342	2045455	chr1	113465200	113465280
  chr1	2045342	2045455	chr1	113465223	113465373
  chr1	2054571	2054420	chr1	45397481	45397526
  chr1	2168591	2168678	chr1	23909210	23909059
  chr1	2796648	2796786	chr1	41629119	41629083
  chr1	3449613	3449763	chr1	187898227	187898378
  chr1	3655493	3655342	chr1	196643496	196643345
  chr1	7753440	7753289	chr1	221959932	221960081
  ~~~

- Code: 

~~~bash
$ sed '1d' GSM6081054_Cell_ID_01.hic.txt |awk '{if ($1 == $4)print $0}' |sort -k1,1 -k2,2n |cut -f 1-6 >Cell_ID_01.bed

# make a loop
$ ls *hic.txt|sed 's/.hic.txt//' |while read line;do sed '1d' $line.hic.txt |awk '{if ($1 == $4)print $0}' |sort -k1,1 -k2,2n |cut -f 1-6 >$line.bed ;done 
~~~





# 2024.3.21

## bedtools intersect

~~~bash
(base) [jialechen@Jiale-MacBook-Pro hic_contact (main âœ—)]$ cut -f 1-3 Cell_ID_01.bed >Cell_ID_01_chrA.bed
(base) [jialechen@Jiale-MacBook-Pro hic_contact (main âœ—)]$ bedtools intersect -a Cell_ID_01_chrA_new.bed -b /Users/jialechen/Desktop/Analysis_work/reference/Hsapiens/ENCFF420VPZ.bed
~~~





 # 2024.3.27 contact motif analysis

## find the motifs regions with hic_contact

~~~bash
# step 1: seperate the bed file to 2 files
$  cut -f 1-3  Cell_ID_12.bed   > Cell_ID_12_1.bed 
$  cut -f 4-6  Cell_ID_12.bed   > Cell_ID_12_2.bed 
# step 2: install genome 
$ sudo perl /Users/jialechen/Desktop/Analysis_work/App/homer/.//configureHomer.pl -install hg38
# step 3: use homer
$ findMotifsGenome.pl Cell_ID_12_1.bed hg38 ../motif/ -size 200 -mask

~~~



# 2024.4.2 Urea condition in BSA condensate

- Analyze the adding the Urea of different concertration to BSA condensate and explore the appropriate condition.

![0321_statis_signi](/Users/jialechen/Desktop/PhD/Condensate/0321_statis_signi.png)

# 2024.4.18

- [Building a custom data service - An example using single-cell Hi-C data](https://github.com/nucleome/Tutorial-SingeCellHiC?tab=readme-ov-file#start_demo)



- [scHiCTools](https://github.com/liu-bioinfo-lab/scHiCTools)

A computational toolbox for analyzing single cell Hi-C (high-throughput sequencing for 3C) data which includes functions for:

1. Loading single-cell HiC datasets
2. Screening valid single-cell data.
3. Smoothing the contact maps with linear convolution, random walk or network enhancing
4. Calculating pairwise similarity using measures include InnerProduct, HiCRep and Selfish
5. Calculating embeddings for single cell HiC datasets efficiently with MDS, t-SNE and PHATE
6. Clustering the cells using scHiCluster, k-means and spectral clustering.
7. Visualizing embedded cells via 2-D or 3-D scatter plot.



# 2024.04.26

- Hic_contact file format transform

~~~bash
java -Xmx48000m  -Djava.awt.headless=true -jar   ~/Desktop/Analysis_work/App/juicer-1.6/scripts/common/juicer_tools.jar pre --threads 16 GSM6081061_Cell_ID_08.hic.txt Cell_ID_08.hic hg38.genome
~~~





